### PR TITLE
Mitigate Command Injection Vulnerability (CVE-2022-45907)

### PR DESCRIPTION
**Pull request by Ma5onic**
Affected versions of this package are vulnerable to Command Injection in `torch.jit.annotations.parse_type_line` which can cause arbitrary code execution because eval is used unsafely. (CVE-2022-45907)

- https://www.cve.org/CVERecord?id=CVE-2022-45907
- https://security.snyk.io/package/pip/torch/1.10.1

- https://github.com/pytorch/pytorch/commit/767f6aa49fe20a2766b9843d01e3b7f7793df6a3

### CODE REVIEW

File: gpu.Dockerfile [hunk 1]
```diff
@@ -29,7 +29,7 @@ RUN mkdir -p /webapp/media /webapp/staticfiles
 
 WORKDIR /webapp
 COPY requirements.txt /webapp/
-RUN pip3 install torch==1.10.2+cu111 -f https://download.pytorch.org/whl/torch_stable.html
+RUN pip3 install torch torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/cu116
```
Comment (JeffreyCA) [inline gpu.Dockerfile:32]: Can you pin the version to `1.13.1`/`0.13.1`? Also we can remove `torchvision` as it's not used.

**Review by JeffreyCA — CHANGES_REQUESTED** [review]
Thanks for the PR!

File: gpu.Dockerfile (hunk 1 above)
Comment (Ma5onic) [reply gpu.Dockerfile:32]: Sure, It has already been pinned in the `requirements.txt`. I'll add it to the dockerfile too.

Comment (Ma5onic) [reply gpu.Dockerfile]: @JeffreyCA, I made the changes that you requested to this branch. 😄 
Let me know if you need anything else

**Review by JeffreyCA — APPROVED** [review]

**Comment by JeffreyCA** [comment]
Hi, unfortunately I had to revert these changes because versions of `torch` greater than 1.10.2 introduce CUDA errors for some GPUs. I tried to see if upgrading the Docker image CUDA version would work but there were conflicts with `nnabla` so 1.10.2 is the latest that can be used w/o any issues.

I didn't see any usages of `torch.jit.annotations.parse_type_line` in dependencies so it may be lower risk

**Comment by Ma5onic** [comment]
[Nvidia's documentation](https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/install-guide.html#linux-distributions) mentions that "Ubuntu 20.04 and Ubuntu 22.04 packages are symlinked (redirected) to ubuntu18.04" and at this point, a working setup can be tested by running a base CUDA container (according to their docs):
```bash
sudo docker run --rm --gpus all nvidia/cuda:11.6.2-base-ubuntu20.04 nvidia-smi
```
I have a feeling that their documentation is a bit out-of-date.

However, I found this useful driver compatibility chart for Nvidia toolkit versions:
https://docs.nvidia.com/cuda/cuda-toolkit-release-notes/index.html#id3
CUDA 11.2.x & CUDA 11.8.x both share the same minimum driver requirements (Linux:`>=450.80.02` | WindowS: `>=452.39`) and the Dockerfiles for cudatoolkit [11.2](https://gitlab.com/nvidia/container-images/cuda/-/blob/master/dist/11.2.0/ubuntu1804/devel/cudnn8/Dockerfile) & [11.8](https://gitlab.com/nvidia/container-images/cuda/-/blob/master/dist/11.8.0/ubuntu2204/devel/cudnn8/Dockerfile) look the same other than the CUDA & OS version numbers. This leads me to believe that we can change `nvidia/cuda:11.2.0-cudnn8-devel-ubuntu20.04` to `nvidia/cuda:11.8.0-cudnn8-devel-ubuntu22.04` in [gpu.Doclerfile](https://github.com/JeffreyCA/spleeter-web/blob/master/gpu.Dockerfile) without introducing breaking changes. (This is just a guess)

nvidia/cuda:11.2.0-cudnn8-devel-ubuntu20.04:
[Dockerfile](https://gitlab.com/nvidia/container-images/cuda/-/blob/master/dist/11.2.0/ubuntu2004/devel/cudnn8/Dockerfile)
[DockerHub](https://hub.docker.com/r/nvidia/cuda/tags?page=1&name=11.2.0-cudnn8-devel-ubuntu20.04)

nvidia/cuda:11.8.0-cudnn8-devel-ubuntu22.04
[Dockerfile](https://gitlab.com/nvidia/container-images/cuda/-/blob/master/dist/11.8.0/ubuntu2204/devel/cudnn8/Dockerfile)
[DockerHub](https://hub.docker.com/r/nvidia/cuda/tags?page=1&name=11.8.0-cudnn8-devel-ubuntu22.04)

Torch will eventually need to be updated/upgraded, I'm hoping that changing to newer cuda versions will help with compatibility.
I took a look at your [requirements.txt](https://github.com/JeffreyCA/spleeter-web/blob/master/requirements.txt) found that `nnabla==1.26.0` and `nnabla-ext-cuda110==1.26.0; sys_platform == 'linux'` were listed, whereas your [requirements.lock](https://github.com/JeffreyCA/spleeter-web/blob/master/requirements.lock) lists `nnabla==1.30.0` and `nnabla-ext-cuda114==1.32.0; sys_platform == 'linux'`.
Changing the requirements  to `nnabla==1.32.1` and `nnabla_ext_cuda114==1.32.1; sys_platform == 'linux'` might help. Using underscores `_` instead of dashes `-` makes you download the cudnn8 compatible version of nnabla (according to their release from 3 days ago: https://github.com/sony/nnabla-ext-cuda/releases/tag/v1.32.1).

Also, which GPUs were problematic?

**Comment by Ma5onic** [comment]
If you could paste the error, that would also be appreciated (if you still have access to it)

**Comment by JeffreyCA** [comment]
It was some memory access error - it was happening with a Tesla K80 which is probably too old. 

I'm thinking of creating a "legacy" compose definition to be able to support older cards and update the current one to what you suggested.

**Comment by Ma5onic** [comment]
I'm all for the idea of legacy GPU support! While you do that, I made a new branch on my fork to test updates, see commit https://github.com/Ma5onic/spleeter-web/commit/7580f1788a49146c1ab34f6c9a562ff7a62ea55b to get an idea of what I am planning for my first upgrade test.
I didn't want to directly upgrade to 22.04 & without testing the upgrade to a newer (11.6.2) cuda toolkit & python 3.9; in hopes that it will still work with the problematic GPU(s)

**Pull request merged**